app/scripts/modules/core/src/securityGroup/SecurityGroups.tsx

```diff
@@ -0,0 +1,157 @@
+  <div>
+    {groups.map(group => (
+      <div key={group.heading} className="rollup">
+        {group.subgroups &&
+          group.subgroups.map(subgroup => (
+            <SecurityGroupPod
+              key={subgroup.heading}
+              grouping={subgroup}
+              application={app}
+              parentHeading={group.heading}
+            />
+          ))}
+      </div>
+    ))}
+    {groups.length === 0 && (
+      <div>
+        <h4 className="text-center">No {FirewallLabels.get('firewalls')} match the filters you've selected.</h4>
+      </div>
+    )}
+  </div>
+);
+
+const Filters = () => {
+  const toggleParam = (event: any): void => {
+    const { checked } = event.target;
+    const name: keyof ISortFilter = event.target.name;
+    (SecurityGroupState.filterModel.asFilterModel.sortFilter[name as keyof ISortFilter] as any) = !!checked;
+    SecurityGroupState.filterModel.asFilterModel.applyParamsToUrl();
+  };
+
+  // The way these "hideX" stateParams work is extra confusing and we should never do this inverse-boolean thing again
+  return (
+    <div className="col-lg-8 col-md-10">
+      <div className="form-inline clearfix filters">
+        <div className="form-group">
+          <label className="checkbox"> Show </label>
+          <div className="checkbox">
+            <label>
+              <input
+                type="checkbox"
+                name="showServerGroups"
+                checked={ReactInjector.$stateParams.hideServerGroups === true}
+                onChange={toggleParam}
+              />{' '}
+              Server Groups
+            </label>
+          </div>
+          <div className="checkbox">
+            <label>
+              <input
+                type="checkbox"
+                name="showLoadBalancers"
+                checked={ReactInjector.$stateParams.hideLoadBalancers === true}
+                onChange={toggleParam}
+              />{' '}
+              Load Balancers
+            </label>
+          </div>
+        </div>
+      </div>
+    </div>
+  );
+};
+
+export const SecurityGroups = ({ app }: ISecurityGroupsProps) => {
+  const [filterModel, setFilterModel] = useState<IFilterModel>({ groups: [], tags: [] });
+  const [initialized, setInitialized] = useState(false);
+
+  const groupsUpdated = () => {
+    setFilterModel({
+      groups: SecurityGroupState.filterModel.asFilterModel.groups,
+      tags: SecurityGroupState.filterModel.asFilterModel.tags,
+    });
+  };
+
+  const updateSecurityGroupGroups = () => {
+    SecurityGroupState.filterModel.asFilterModel.applyParamsToUrl();
+    // If we are using managed resources, wait until they are ready before updating security groups.
+    // Otherwise, the managed resource fields will not be present on the security group groupings and we'll lose the
+    // badges on the group headers
+    const waiter = SETTINGS.feature.managedResources ? app.managedResources.ready() : Promise.resolve();
```

Comment:
@erikmunson check this out lol the debounce was masking this (`updateSecurityGroups` gets called when the `securityGroups` dataSource refreshes, which happens a tick before the managed resources get added.

There are separate questions this raises about the composability of data sources and their lifecycles but I'm not getting into that.

Reply:
incredible. Ever since I looked into the joins that we do between the entity tag data source (and subsequently copied it wholesale), I've wanted to write some kind of first-class composition thing that's part of the data source API and does magic with [`combineLatest`](https://www.learnrxjs.io/operators/combination/combinelatest.html) to join stuff in a way that hooks up to the refresh lifecycle, this makes me feel like it's even more actually a thing that should happen.